File: bestfitr/tests/testthat/test-sobol.R
# Identity test for the Sobol quasi-random sequence.
# Pins the R binding to the canonical first-10 points for dimension 2.
# Oracle values come from upstream/Numerics/Test_Numerics/Sampling/Test_SobolSequence.cs
# (Test_Sobol), which are themselves verified against R's randtoolbox::sobol().

ns <- asNamespace("bestfitr")

# Locate the direction-numbers file installed with the package.
path <- system.file("extdata", "new-joe-kuo-6.21201", package = "bestfitr")
if (!nzchar(path)) stop("direction-numbers file not found in bestfitr inst/extdata")

test_that("sobol dimension-2 first 10 points match C# oracle", {
  mat <- ns$bf_sobol_generate_(2L, 10L, path)
  expected <- rbind(
    c(0.5000, 0.5000),
    c(0.7500, 0.2500),
    c(0.2500, 0.7500),
    c(0.3750, 0.3750),
    c(0.8750, 0.8750),
    c(0.6250, 0.1250),
    c(0.1250, 0.6250),
    c(0.1875, 0.3125),
    c(0.6875, 0.8125),
    c(0.9375, 0.0625)
  )
  for (i in seq_len(nrow(expected))) {
    for (j in seq_len(ncol(expected))) {
      expect_equal(mat[i, j], expected[i, j], tolerance = 0,
                   label = paste0("point[", i, ",", j, "]"))
    }
  }
})

test_that("sobol dimension-1 first point is 0.5", {
  # Dimension 1 needs no file (unit initialization only).
  mat <- ns$bf_sobol_generate_(1L, 1L, "")
  expect_equal(mat[1, 1], 0.5, tolerance = 0)
})

test_that("sobol skip_to(1) matches first sequential point", {
  first_seq <- ns$bf_sobol_generate_(2L, 1L, path)[1, ]
  first_skip <- ns$bf_sobol_skip_to_(2L, 1L, path)
  expect_equal(first_skip, first_seq, tolerance = 0)
})

test_that("sobol skip_to(5) matches 5th sequential point", {
  fifth_seq <- ns$bf_sobol_generate_(2L, 5L, path)[5, ]
  fifth_skip <- ns$bf_sobol_skip_to_(2L, 5L, path)
  expect_equal(fifth_skip, fifth_seq, tolerance = 0)
})

File: corehydropy/tests/test_sobol.py
"""Identity tests for the Sobol quasi-random sequence.

Pins the Python binding to the canonical first-10 points for dimension 2.
Oracle values come from upstream/Numerics/Test_Numerics/Sampling/Test_SobolSequence.cs
(Test_Sobol), verified against R's randtoolbox::sobol().
"""
from __future__ import annotations

from importlib.resources import files

import pytest

from corehydropy import _core


def _data_path() -> str:
    """Return the path to the installed direction-numbers file."""
    return str(files("corehydropy") / "data" / "new-joe-kuo-6.21201")


# Oracle: first 10 points for dimension 2 (exact binary fractions).
_EXPECTED_DIM2 = [
    [0.5000, 0.5000],
    [0.7500, 0.2500],
    [0.2500, 0.7500],
    [0.3750, 0.3750],
    [0.8750, 0.8750],
    [0.6250, 0.1250],
    [0.1250, 0.6250],
    [0.1875, 0.3125],
    [0.6875, 0.8125],
    [0.9375, 0.0625],
]


def test_sobol_dim2_first10():
    """First 10 points for dimension 2 match the C# oracle (exact equality)."""
    path = _data_path()
    pts = _core.sobol_generate(2, 10, path)
    for i, (actual, expected) in enumerate(zip(pts, _EXPECTED_DIM2)):
        for j in range(2):
            assert actual[j] == expected[j], (
                f"point[{i}][{j}]: got {actual[j]}, expected {expected[j]}"
            )


def test_sobol_dim1_first_point():
    """Dimension 1 first point is 0.5 (unit initialization, no file needed)."""
    pts = _core.sobol_generate(1, 1, "")
    assert pts[0][0] == 0.5


def test_sobol_skip_to_first():
    """skip_to(1) returns the same value as the first sequential point."""
    path = _data_path()
    first_seq = _core.sobol_generate(2, 1, path)[0]
    first_skip = _core.sobol_skip_to(2, 1, path)
    for j in range(2):
        assert first_skip[j] == first_seq[j]


def test_sobol_skip_to_fifth():
    """skip_to(5) matches the 5th sequential point (0-indexed: pts[4])."""
    path = _data_path()
    fifth_seq = _core.sobol_generate(2, 5, path)[4]
    fifth_skip = _core.sobol_skip_to(2, 5, path)
    for j in range(2):
        assert fifth_skip[j] == fifth_seq[j]


def test_sobol_identity_r_python():
    """Python sequence matches the R-validated oracle values bit-for-bit."""
    path = _data_path()
    pts = _core.sobol_generate(2, 10, path)
    for i, expected_row in enumerate(_EXPECTED_DIM2):
        for j, expected_val in enumerate(expected_row):
            assert pts[i][j] == expected_val, (
                f"identity mismatch at point[{i}][{j}]"
            )
